File: src/server/qdrant.rs
use qdrant_client::qdrant::{
    CreateCollectionBuilder, Distance, Filter, PointStruct,
    ScalarQuantizationBuilder, VectorParamsBuilder,
    CreateFieldIndexCollectionBuilder, FieldType,
    SearchPointsBuilder, UpsertPointsBuilder, DeletePointsBuilder,
    Condition, PointId, points_selector::PointsSelectorOneOf,
};
use qdrant_client::Qdrant;
use serde_json::json;
use uuid::Uuid;

pub const COLLECTION_NAME: &str = "book_chunks";
const VECTOR_SIZE: u64 = 1024;

pub async fn ensure_collection(client: &Qdrant) -> Result<(), qdrant_client::QdrantError> {
    let exists = client.collection_exists(COLLECTION_NAME).await?;
    if exists {
        return Ok(());
    }

    client
        .create_collection(
            CreateCollectionBuilder::new(COLLECTION_NAME)
                .vectors_config(VectorParamsBuilder::new(VECTOR_SIZE, Distance::Cosine))
                .quantization_config(ScalarQuantizationBuilder::default()),
        )
        .await?;

    client
        .create_field_index(
            CreateFieldIndexCollectionBuilder::new(COLLECTION_NAME, "book_id", FieldType::Keyword),
        )
        .await?;
    client
        .create_field_index(
            CreateFieldIndexCollectionBuilder::new(COLLECTION_NAME, "tags", FieldType::Keyword),
        )
        .await?;
    client
        .create_field_index(
            CreateFieldIndexCollectionBuilder::new(COLLECTION_NAME, "authors", FieldType::Keyword),
        )
        .await?;

    Ok(())
}

pub async fn delete_book_points(
    client: &Qdrant,
    book_id: Uuid,
) -> Result<(), qdrant_client::QdrantError> {
    let filter = Filter::must([Condition::matches(
        "book_id",
        book_id.to_string(),
    )]);

    client
        .delete_points(
            DeletePointsBuilder::new(COLLECTION_NAME)
                .points(PointsSelectorOneOf::Filter(filter)),
        )
        .await?;

    Ok(())
}

pub async fn upsert_chunks(
    client: &Qdrant,
    book_id: Uuid,
    ref_id: &str,
    title: &str,
    authors: &[String],
    tags: &[String],
    chunks: &[(usize, Option<String>, usize, String)],
    embeddings: &[Vec<f32>],
) -> Result<(), qdrant_client::QdrantError> {
    let mut points = Vec::new();

    for ((chapter_idx, chapter_title, chunk_index, text), embedding) in
        chunks.iter().zip(embeddings.iter())
    {
        let point_id = Uuid::new_v4();
        let payload = json!({
            "book_id": book_id.to_string(),
            "ref_id": ref_id,
            "title": title,
            "chunk_index": chunk_index,
            "chunk_text": text,
            "chapter_idx": chapter_idx,
            "chapter": chapter_title.as_deref().unwrap_or(""),
            "authors": authors,
            "tags": tags,
        });

        let payload_map: std::collections::HashMap<String, qdrant_client::qdrant::Value> =
            serde_json::from_value(payload).unwrap_or_default();

        points.push(PointStruct::new(
            PointId::from(point_id.to_string()),
            embedding.clone(),
            payload_map,
        ));

        if points.len() >= 100 {
            let batch: Vec<PointStruct> = points.drain(..).collect();
            client
                .upsert_points(
                    UpsertPointsBuilder::new(COLLECTION_NAME, batch),
                )
                .await?;
        }
    }

    if !points.is_empty() {
        client
            .upsert_points(
                UpsertPointsBuilder::new(COLLECTION_NAME, points),
            )
            .await?;
    }

    Ok(())
}

pub struct SearchResult {
    pub ref_id: String,
    pub title: String,
    pub chunk_text: String,
    pub score: f32,
}

pub async fn search_similar(
    client: &Qdrant,
    query_embedding: Vec<f32>,
    tags: &[String],
    limit: u64,
) -> Result<Vec<SearchResult>, qdrant_client::QdrantError> {
    let mut search = SearchPointsBuilder::new(COLLECTION_NAME, query_embedding, limit)
        .with_payload(true);

    if !tags.is_empty() {
        let filter = Filter::must(
            tags.iter()
                .map(|t| Condition::matches("tags", t.clone()))
                .collect::<Vec<_>>(),
        );
        search = search.filter(filter);
    }

    let results = client.search_points(search).await?;

    let mut search_results = Vec::new();
    for point in results.result {
        let payload = &point.payload;
        let ref_id = payload
            .get("ref_id")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string())
            .unwrap_or_default();
        let title = payload
            .get("title")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string())
            .unwrap_or_default();
        let chunk_text = payload
            .get("chunk_text")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string())
            .unwrap_or_default();

        search_results.push(SearchResult {
            ref_id,
            title,
            chunk_text,
            score: point.score,
        });
    }

    Ok(search_results)
}

File: src/lib/qdrant.ts
import { QdrantClient } from "@qdrant/js-client-rest";

const COLLECTION_NAME = "book_chunks";
const VECTOR_SIZE = 1024;

let client: QdrantClient;

export function getQdrant(): QdrantClient {
  if (!client) {
    client = new QdrantClient({
      url: process.env.QDRANT_URL!,
      apiKey: process.env.QDRANT_API_KEY || undefined,
    });
  }
  return client;
}

export async function ensureCollection(): Promise<void> {
  const q = getQdrant();
  const exists = await q.collectionExists(COLLECTION_NAME);
  if (exists.exists) return;

  await q.createCollection(COLLECTION_NAME, {
    vectors: { size: VECTOR_SIZE, distance: "Cosine" },
    quantization_config: {
      scalar: { type: "int8", always_ram: true },
    },
  });

  await q.createPayloadIndex(COLLECTION_NAME, {
    field_name: "book_id",
    field_schema: "keyword",
  });
  await q.createPayloadIndex(COLLECTION_NAME, {
    field_name: "tags",
    field_schema: "keyword",
  });
  await q.createPayloadIndex(COLLECTION_NAME, {
    field_name: "authors",
    field_schema: "keyword",
  });
}

export async function deleteBookPoints(bookId: string): Promise<void> {
  const q = getQdrant();
  await q.delete(COLLECTION_NAME, {
    filter: {
      must: [{ key: "book_id", match: { value: bookId } }],
    },
  });
}

export interface QdrantChunk {
  chapterIdx: number;
  chapterTitle: string | null;
  chunkIndex: number;
  text: string;
}

export async function upsertChunks(
  bookId: string,
  reference: string,
  title: string,
  authors: string[],
  tags: string[],
  chunks: QdrantChunk[],
  embeddings: number[][],
): Promise<void> {
  const q = getQdrant();
  const points: {
    id: string;
    vector: number[];
    payload: Record<string, unknown>;
  }[] = [];

  for (let i = 0; i < chunks.length; i++) {
    const chunk = chunks[i];
    const embedding = embeddings[i];
    points.push({
      id: crypto.randomUUID(),
      vector: embedding,
      payload: {
        book_id: bookId,
        reference,
        title,
        chunk_index: chunk.chunkIndex,
        chunk_text: chunk.text,
        chapter_idx: chunk.chapterIdx,
        chapter: chunk.chapterTitle ?? "",
        authors,
        tags,
      },
    });

    if (points.length >= 100) {
      const batch = points.splice(0, points.length);
      await q.upsert(COLLECTION_NAME, { points: batch });
    }
  }

  if (points.length > 0) {
    await q.upsert(COLLECTION_NAME, { points });
  }
}

export interface SearchResult {
  reference: string;
  title: string;
  chunk_text: string;
  score: number;
}

export async function searchSimilar(
  queryEmbedding: number[],
  tags: string[],
  author: string | null,
  limit: number,
): Promise<SearchResult[]> {
  const q = getQdrant();

  const must: Array<{
    key: string;
    match: { value: string };
  }> = [];

  for (const t of tags) {
    must.push({ key: "tags", match: { value: t } });
  }
  if (author) {
    must.push({ key: "authors", match: { value: author } });
  }

  const results = await q.search(COLLECTION_NAME, {
    vector: queryEmbedding,
    limit,
    with_payload: true,
    ...(must.length > 0 ? { filter: { must } } : {}),
  });

  return results.map((point) => ({
    reference: (point.payload?.reference as string) ?? "",
    title: (point.payload?.title as string) ?? "",
    chunk_text: (point.payload?.chunk_text as string) ?? "",
    score: point.score,
  }));
}
